source/Nevermore/Advanced/ReadTransaction.cs

```diff
@@ -905,22 +905,27 @@ DbTransaction BeginTransaction(IsolationLevel isolationLevel, string sqlServerTr
 
         public void Dispose()
         {
-            // ReSharper disable ConstantConditionalAccessQualifier
-            if (OwnsSqlTransaction)
+            try
             {
-                Transaction?.Dispose();
-            }
-
-            TransactionTimer?.Dispose();
-            DeadlockAwareLock?.Dispose();
+                // ReSharper disable ConstantConditionalAccessQualifier
+                if (OwnsSqlTransaction)
+                {
+                    Transaction?.Dispose();
+                }
 
-            if (OwnsSqlTransaction)
-            {
-                connection?.Dispose();
+                TransactionTimer?.Dispose();
+                DeadlockAwareLock?.Dispose();
             }
+            finally
+            {
+                if (OwnsSqlTransaction)
+                {
+                    connection?.Dispose();
+                }
 
-            // ReSharper restore ConstantConditionalAccessQualifier
-            registry.Remove(this);
+                // ReSharper restore ConstantConditionalAccessQualifier
+                registry.Remove(this);
```

Comment:
If Dispose throws, the transaction will remain in the registry, and will probably be a memory leak.
We should probably put some safety code around this, or perhaps remove it before calling other code that might throw.

I'd be interested in some other opinions here though, particularly @rosslovas and @adam-mccoy who know this code better than I. 

I "soft-approve" the PR, but will wait on final approval for a second opinion

Reply:
I thought had a comment here but it seems GitHub lost it. I was going to ask how we should deal with this.
It's essentially a HashSet.Remove(), and I'm not concerned about it throwing. Maybe we should move this to be first in the `finally` block?

Reply:
Yeah I was thinking that too. Let's see what Ross or Adam thinks

Reply:
I left a top-level comment that might make worrying about this unnecessary, but if you don't go that direction then I agree; it's hard to imagine this throwing so putting it first sounds fine 🙂